File: life.py
#!/usr/bin/env python3
import time

# Rules
D = 0  # Die
G = 1  # Grow
S = 2  # Stay the same


class Life:
    def __init__(self, width=10, height=10, rules=[D, D, S, G, D, D, D, D, D]):
        self.generation = 0
        self.width = width
        self.height = height
        self.rules = rules
        self.grid = [[False for i in range(height)] for j in range(width)]

    def get(self, x, y):
        # No wrapping
        if 0 <= x < self.width and 0 <= y < self.height:
            return self.grid[x][y]
        return 0

    def get_neighbors(self, x, y):
        neighbors = 0
        for x_offset in [-1, 0, 1]:
            for y_offset in [-1, 0, 1]:
                # Don't count the offset 0, 0
                if not x_offset == y_offset == 0:
                    if self.get(x+x_offset, y+y_offset):
                        neighbors += 1
        return neighbors

    def print_grid(self):
        print(f'Generation {self.generation}')
        print('+{}+'.format('-'*self.width))
        for row in self.grid:
            row_text = ''.join(['0' if x else ' ' for x in row])
            print(f'|{row_text}|')
        print('+{}+'.format('-'*self.width))

    def next_generation(self):
        next_grid = [y[:] for y in self.grid[:]]
        for x in range(self.width):
            for y in range(self.height):
                neighbors = self.get_neighbors(x, y)
                next_state = self.rules[neighbors]
                if next_state == G:
                    next_grid[x][y] = True
                elif next_state == D:
                    next_grid[x][y] = False
        self.grid = next_grid
        self.generation += 1

    def add_figure(self, points, x=5, y=5):
        for a, b in points:
            self.grid[x+a][y+b] = True


RPENTOMINO = [[1, 0], [2, 0], [0, 1], [1, 1], [1, 2]]
BLOCK = [[0, 0], [1, 0], [0, 1], [1, 1]]
BLINKER = [[1, 0], [1, 1], [1, 2]]
BEACON = [[0, 0], [0, 1], [1, 0], [2, 3], [3, 2], [3, 3]]

if __name__ == '__main__':
    game = Life()
    # game.add_figure(BEACON, x=2, y=1)
    game.add_figure(RPENTOMINO, x=4, y=4)

    while True:
        game.print_grid()
        game.next_generation()
        time.sleep(1)

File: life.pl
#!/usr/bin/env perl

#Grid size
$WIDTH = 20;
$HEIGHT = 10;

# Rule per number of neighbors
use constant {
    D => 0,  # Die
    S => 1,  # Stay the same
    G => 2,  # Grow
};
#      0 1 2 3 4 5 6 7 8
@RULES = (D, D, S, G, D, D, D, D, D);

my @Grid;

sub get_index {
    my ($x, $y) = @_;
    return $x*$WIDTH + $y;

}

sub get {
    my ($x, $y) = @_;
    if ($x >= 0 && $x < $WIDTH && $y >= 0 && $y < $HEIGHT) {
        return $Grid[get_index($x, $y)];
    }
}

sub get_neighbors {
    my ($x, $y) = @_;
    my $neighbors = 0;
    for my $x_offset (-1 .. 1) {
        for my $y_offset (-1 .. 1) {
            # Don't count the offset 0, 0
            if ($x_offset != $y_offset or $x_offset != 0) {
                if (get($x+$x_offset, $y+$y_offset)) {
                    $neighbors += 1
                }
            }
        }
    }
    return $neighbors
}

sub print_grid {
    print("+" . "-" x $WIDTH . "+\n");
    for my $y (0 .. $HEIGHT-1) {
        print("|");
        for my $x (0 .. $WIDTH-1) {
            if (get($x, $y)) {
                print("0");
            } else {
                print(" ");
            }
        }
        print("|\n");
    }
    print("+" . "-" x $WIDTH . "+\n");
}

sub next_generation {
    my @next_grid;
    for my $x (0 .. $WIDTH-1) {
        for my $y (0 .. $HEIGHT-1) {
            my $neighbors = get_neighbors($x, $y);
            my $next_state = $RULES[$neighbors];
            if ($next_state == G) {
                $next_grid[get_index($x, $y)] = 1;
            } elsif ($next_state == S) {
                $next_grid[get_index($x, $y)] = $Grid[get_index($x, $y)];
            }
        }
    }
    for my $x (0 .. $WIDTH-1) {
        for my $y (0 .. $HEIGHT-1) {
            $Grid[get_index($x, $y)] = $next_grid[get_index($x, $y)];
        }
    }
}

sub add_figure {
    my $x = shift;
    my $y = shift;
    while (@_) {
        my $a = shift;
        my $b = shift;
        $Grid[get_index($x+$a, $y+$b)] = 1;
    }
}

@RPENTOMINO = ((1, 0), (2, 0), (0, 1), (1, 1), (1, 2));
@BLOCK = ((0, 0), (1, 0), (0, 1), (1, 1));
@BLINKER = ((1, 0), (1, 1), (1, 2));
@BEACON = ((0, 0), (0, 1), (1, 0), (2, 3), (3, 2), (3, 3));

# add_figure(0, 0, @BLINKER);
add_figure(4, 4, @RPENTOMINO);

my $generation = 0;
while (1) {
    print("Generation $generation\n");
    print_grid();
    next_generation();
    $generation += 1;
    sleep(1)
}
